File: Create/Singleton/Singleton.py
import threading

class PrinterSpooler:
    _instance = None
    _lock = threading.Lock()

    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    print("PrinterSpooler instantiated")
        return cls._instance

    # 业务方法示例
    def add_job(self, job: str):
        print(f"Queued: {job}")

# Demo
if __name__ == "__main__":
    s1 = PrinterSpooler()
    s2 = PrinterSpooler()
    print(s1 is s2)  # True
    s1.add_job("report.pdf")

File: Create/Singleton/Singleton.java
public final class PrinterSpooler {
    // 1. 静态变量持有唯一实例
    private static volatile PrinterSpooler instance;

    // 2. 私有构造器防止外部 new
    private PrinterSpooler() {
        System.out.println("PrinterSpooler instantiated");
    }

    // 3. 全局访问点（双重检查锁定，线程安全）
    public static PrinterSpooler getInstance() {
        if (instance == null) {
            synchronized (PrinterSpooler.class) {
                if (instance == null) {
                    instance = new PrinterSpooler();
                }
            }
        }
        return instance;
    }

    // 业务方法示例
    public void addJob(String job) {
        System.out.println("Queued: " + job);
    }

    // Demo
    public static void main(String[] args) {
        PrinterSpooler spooler1 = PrinterSpooler.getInstance();
        PrinterSpooler spooler2 = PrinterSpooler.getInstance();
        System.out.println(spooler1 == spooler2); // true
        spooler1.addJob("report.pdf");
    }
}